tools/generator/dfg/sam/avr_device_tree.py

```diff
@@ -0,0 +1,261 @@
+# -*- coding: utf-8 -*-
+# Copyright (c) 2013-2016, Niklas Hauser
+# Copyright (c)      2016, Fabian Greif
+# All rights reserved.
+
+import math
+import logging
+
+from ..device_tree import DeviceTree
+from ..input.xml import XMLReader
+
+from .avr_identifier import AVRIdentifier
+from . import avr_io
+from . import avr_mcu
+
+LOGGER = logging.getLogger('dfg.avr.reader')
+
+class AVRDeviceTree:
+    """ AVRDeviceTree
+    This AVR specific part description file reader knows the structure and
+    translates the data into a platform independent format.
+    """
+
+    @staticmethod
+    def _properties_from_file(filename):
+        device_file = XMLReader(filename)
+
+        device = device_file.query("//device")[0]
```

Comment:
So the way this works is you get data from the vendor XML using XPath queries using etree Python interface.

You dump this into a dictionary of properties (=`p`) and this is like a vendor-specific internal interface that you can change as you wish. You can use pass this dictionary to other functions and modify it, BUT it only exists for one unique device.

Unfortunately there aren't any helper tools for *exploring* the vendor data set, you'll just have to look at a lot of XML. See also: https://blog.salkinium.com/modm-devices/#peeking-into-stm32cubemx